slack/models.py

```diff
@@ -0,0 +1,142 @@
+import datetime
+
+from django.db import models, connection
+
+"""
+`ts` fields from slack are actually IDs that can be interpreted into
+timestamps.  We store them in a string as we get them from slack instead
+of a DateTimeField to be able to pass them back exactly to slack without
+precision or something causing round-tripping errors.
+"""
+
+
+def parse_ts(ts):
+    seconds = float(ts)
+    return datetime.datetime.fromtimestamp(seconds, tz=datetime.timezone.utc)
+
+
+class SlackUser(models.Model):
+    id = models.CharField(max_length=16, primary_key=True)
+    name = models.TextField()
+    real_name = models.TextField()
+    email = models.TextField()
+    image_48 = models.URLField()
+
+
+class Channel(models.Model):
+    id = models.CharField(max_length=16, primary_key=True)
+    name = models.TextField()
+    topic = models.TextField()
+    purpose = models.TextField()
+    # We only need to check for new messages strictly newer than this.
+    last_update_ts = models.CharField(max_length=32, null=True)
+
+
+class ChannelUpdateGap(models.Model):
+    """
+    Stores a (potentially unbounded) period of time where we haven't counted
+    messages for a channel. Used to store progress and allow resumption.
+    """
+
+    channel = models.ForeignKey(Channel, on_delete=models.CASCADE)
+    # There may be more messages to fetch that are strictly older than
+    # oldest_message_ts (null meaning no limit)
+    oldest_message_ts = models.CharField(max_length=32, null=True)
+    # but not strictly newer than newest_message_ts (null meaning no limit)
+    newest_message_ts = models.CharField(max_length=32, null=True)
+
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                check=models.lookups.GreaterThan(
+                    models.functions.Cast(
+                        "newest_message_ts", output_field=models.FloatField()
+                    ),
+                    models.functions.Cast(
+                        "oldest_message_ts", output_field=models.FloatField()
+                    ),
+                ),
+                name="newest_newer_than_oldest",
+            )
+        ]
+
+    def __str__(self):
+        return "(%s, %s)" % (self.oldest_message_ts, self.newest_message_ts)
```

Comment:
I recommend adding a `repr` that includes the class name and id for debugging when overriding the default Django model `str` method like this, since otherwise it can be very annoying to figure out what's what.

```suggestion
        return f"({self.oldest_message_ts}, {self.newest_message_ts})"
    
    def __repr__(self):
        return f"<ChannelUpdateGap: {self} ({self.id})>"
```